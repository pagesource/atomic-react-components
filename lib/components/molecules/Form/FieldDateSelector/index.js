// @flow
export { default } from './FieldDateSelector';
export { FieldDateSelectorVanilla } from './FieldDateSelector';
