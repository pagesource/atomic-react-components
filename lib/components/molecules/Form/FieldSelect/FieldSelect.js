// @flow
import React from 'react';
import type { Node, ComponentType } from 'react';
import styled from 'styled-components';
import Label from '../../../atoms/Label';
import Select from '../../../atoms/Select';
import FieldError from '../../../atoms/FieldError';
import styles from './FieldSelect.style';

type Props = {
  id: string,
  className: string,
  name: string,
  label: Node,
  register: Function,
  errors: Object,
  labelProps?: Object,
  fieldProps?: Object,
  onChange?: Function,
};

const FieldSelect = ({
  id,
  className,
  name,
  label,
  onChange,
  labelProps,
  fieldProps,
  register,
  errors,
  ...others
}: Props): Node => (
  <div className={className}>
    <Label htmlFor={id} {...labelProps}>
      {label}
    </Label>
    <Select {...others} ref={register} name={name} />
    {errors[name] && <FieldError>{errors[name].message}</FieldError>}
  </div>
);

FieldSelect.defaultProps = {
  labelProps: {},
  fieldProps: {},
  onChange: null,
};

const StyledFieldSelect: ComponentType<Props> = styled(FieldSelect)`
  ${styles};
`;

export default StyledFieldSelect;

export { FieldSelect as FieldSelectVanilla };
