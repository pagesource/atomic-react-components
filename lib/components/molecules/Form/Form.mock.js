import { action } from '@storybook/addon-actions';

const userFormProps = {
  initialValues: {
    email: '',
    username: '',
    typeOfUser: 'Admin',
    comment: '',
    ageGroup: '',
    termsAndConditions: false,
  },
  debug: false,
  action: '/',
  validationSchema: {
    username: {
      max: [15, 'Max limit reached'],
      required: 'User name Required',
    },
    email: {
      max: [30, 'Max reached'],
      email: 'Invalid email format',
      required: 'Required',
    },
    comment: {
      max: [100, 'Comment should not be more than 100 characters'],
      required: 'Please enter a comment',
    },
    ageGroup: {
      required: 'Please select an Age group',
    },
    termsAndConditions: {
      type: 'bool',
      test: [
        'termsAndConditions',
        'please accept terms and conditions!!!',
        (value) => value === true,
      ],
    },
  },
  onSubmit: action('On Submit Fired'),
};

const userFormSelectProps = {
  options: ['Admin', 'Read Only'],
  name: 'typeOfUser',
  label: 'Type of user',
};

const userFormNameInputProps = {
  name: 'username',
  type: 'text',
  label: 'User Name',
  validationInstructions: 'Enter valid username',
};

const fileMultipleInputProps = {
  name: 'fileuploadMulti',
  localeText: {
    btnTextPrimary: 'Choose Files',
    btnTextSecondary: 'Upload...',
    dragHelperText: 'Drop Files Here...',
  },
  multiple: true,
};
const fileSingleInputProps = {
  name: 'fileupload',
  localeText: {
    btnTextPrimary: 'Choose File',
    btnTextSecondary: 'Upload...',
    dragHelperText: 'Drop File Here...',
  },
};
const fileDDInputProps = {
  name: 'fileuploadDD',
  localeText: {
    btnTextPrimary: 'Choose File',
    btnTextSecondary: 'Upload...',
    dragHelperText: 'Drop File Here...',
  },
};

const userFormEmailProps = {
  type: 'email',
  name: 'email',
  label: 'Email',
  validationInstructions: 'Enter valid email',
};

const userFormCommentProps = {
  name: 'comment',
  label: 'Comment',
  placeholder: 'Enter your comment here...',
};

const userFormAgeGroupProps = {
  name: 'ageGroup',
  radioGroupArray: [
    {
      value: '30-',
      label: '< 30',
    },
    {
      value: '30-60',
      label: '30-60',
    },
    {
      value: '60+',
      label: '> 60',
    },
  ],
};

const userFormTnCProps = {
  name: 'termsAndConditions',
  label: 'I accept terms & conditions',
};

const userFormToggleProps = {
  name: 'test_1',
  id: 'checkbox_test_1',
  value: 'off',
  label: 'Toggle',
  indicatorOn: 'On',
  indicatorOff: 'Off',
  disabled: false,
  intialToggleState: false,
  displayIndicators: true,
};

export {
  userFormProps,
  fileMultipleInputProps,
  fileSingleInputProps,
  userFormSelectProps,
  userFormNameInputProps,
  userFormEmailProps,
  userFormCommentProps,
  userFormAgeGroupProps,
  userFormTnCProps,
  userFormToggleProps,
  fileDDInputProps,
};
