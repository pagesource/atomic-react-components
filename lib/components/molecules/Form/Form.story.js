// @flow
import React from 'react';
import { storiesOf } from '@storybook/react';
import {
  userFormProps,
  userFormSelectProps,
  fileMultipleInputProps,
  fileSingleInputProps,
  userFormNameInputProps,
  userFormEmailProps,
  userFormCommentProps,
  userFormAgeGroupProps,
  userFormDateSelectorProps,
  userFormTnCProps,
  userFormToggleProps,
} from './Form.mock';

// Import Vanilla Component to display markup and props
import Form from './Form';
import Button from '../../atoms/Button';
// Import Styled Component to showcase variations

storiesOf('Molecules/Form', module)
  .addParameters({ jest: ['Form', 'Button'] })
  .add('Default', () => (
    <Form {...userFormProps}>
      <div>
        <Form.Select {...userFormSelectProps} />
        <Form.Input {...userFormNameInputProps} />
        <Form.Input {...userFormEmailProps} />
        <Form.FieldFile {...fileMultipleInputProps} />
        <Form.FieldFile {...fileSingleInputProps} />
        <Form.DDFile {...fileSingleInputProps} />
        <Form.RadioGroup {...userFormAgeGroupProps} />
        <Form.DateSelector {...userFormDateSelectorProps} />
        <Form.Checkbox {...userFormTnCProps} />
        <Form.Textarea {...userFormCommentProps} />
        <Form.Toggle {...userFormToggleProps} />
      </div>
      <Button primary type="submit">
        Submit
      </Button>
    </Form>
  ));
