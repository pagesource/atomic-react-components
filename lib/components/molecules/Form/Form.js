// @flow
import React from 'react';
import { useForm } from 'react-hook-form';
import type { Node } from 'react';
import { buildSchema } from '../../../core/utils/buildValidationSchema';
import Debug from './Debug';
import FieldInput from './FieldInput';
import FieldSelect from './FieldSelect';
import FieldTextarea from './FieldTextarea';
import FieldRadioGroup from './FieldRadioGroup';
import FieldCheckbox from './FieldCheckbox';
import FieldToggle from './FieldToggle';
import FieldFileInput from './FieldFileInput';
import DDFileInput from './DragDropInput';

type Props = {
  initialValues?: any,
  children: Node,
  onSubmit: Function,
  debug?: boolean,
  validationSchema?: Object | null,
  validate?: String,
  yupValidationSchema?: Function | null,
};

const Form = ({
  onSubmit,
  initialValues,
  children,
  debug,
  validate,
  validationSchema,
  yupValidationSchema,
}: Props) => {
  const getSchema = () => {
    if (yupValidationSchema) return yupValidationSchema;
    return validationSchema ? buildSchema(validationSchema) : null;
  };
  const { handleSubmit, register, errors, setValue } = useForm({
    mode: validate,
    validationSchema: getSchema(),
    defaultValues: initialValues,
  });

  // $flow-disable-line  eslint-disable-line
  const childrenWithProps = children[0].props.children.map((child) => {
    return React.cloneElement(child, { register, errors, setValue });
  });

  // $flow-disable-line  eslint-disable-line
  childrenWithProps.push(children[1]);
  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      {childrenWithProps}
      {debug && <Debug />}
    </form>
  );
};

Form.Checkbox = FieldCheckbox;

Form.Input = FieldInput;

Form.RadioGroup = FieldRadioGroup;

Form.Select = FieldSelect;

Form.Textarea = FieldTextarea;

Form.Toggle = FieldToggle;

Form.FieldFile = FieldFileInput;

Form.DDFile = DDFileInput;

Form.defaultProps = {
  debug: false,
  initialValues: null,
  validationSchema: null,
  yupValidationSchema: null,
  validate: 'onSubmit',
};

export default Form;
