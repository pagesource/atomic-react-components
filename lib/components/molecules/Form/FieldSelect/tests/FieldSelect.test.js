import React from 'react';
import {
  shallowWithTheme,
  mountWithTheme,
  renderWithTheme,
} from '../../../../../../__mock__/themeMock';
import { FieldSelectVanilla } from '../FieldSelect';
import Form from '../../Form';

describe('FieldInput Component', () => {
  let FieldInput = '';

  test('should render correctly', () => {
    FieldInput = shallowWithTheme(
      <Form>
        <FieldSelectVanilla />
      </Form>
    );
    expect(FieldInput).toMatchSnapshot();
  });

  test('should render label correctly', () => {
    const props = {
      label: 'Select Fav Color',
      options: ['red', 'black', 'blue'],
      name: 'color',
    };
    FieldInput = renderWithTheme(
      <Form>
        <FieldSelectVanilla {...props} />
      </Form>
    );
    expect(FieldInput.find('label').text()).toEqual('Select Fav Color');
  });

  test('should render select tag', () => {
    const props = {
      label: 'Select Fav Color',
      options: ['red', 'black', 'blue'],
      name: 'color',
    };
    FieldInput = mountWithTheme(
      <Form>
        <FieldSelectVanilla {...props} />
      </Form>
    );
    expect(FieldInput.find('select')).toHaveLength(1);
  });
});
