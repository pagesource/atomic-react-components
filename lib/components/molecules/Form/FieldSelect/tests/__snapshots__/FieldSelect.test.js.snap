// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldInput Component should render correctly 1`] = `
<Formik
  initialValues={null}
  validate={[Function]}
  validateOnBlur={true}
  validateOnChange={true}
  validationSchema={null}
>
  <Component />
</Formik>
`;
